File: neetcode150/medium/39-Combination-Sum.py
# Solution 1
class Solution:
    def combinationSum(self, candidates, target):
        res = []
        curr = []

        def createSum(nums, target, index):
            if target == 0:
                res.append(curr.copy())
                return 
            for i in range(index, len(nums)):
                if target - nums[i] >= 0:
                    curr.append(nums[i])
                    createSum(nums, target - nums[i], i)
                    curr.pop()
        createSum(candidates, target, 0)
        return res    

File: neetcode150/medium/39-Combination-Sum.cpp
#include <vector>
using namespace std;

// Solution 1
class Solution {
public:
    vector<vector<int>> res;
    vector<int> curr; 
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        createSum(candidates, target, 0);
        return res;
    }
    void createSum(vector<int>& nums, int target, int index){
        if(target == 0){
            res.push_back(curr);
            return;
        }
        for(int i = index; i < nums.size(); i++){
            if(target - nums[i] >= 0){
                curr.push_back(nums[i]);
                createSum(nums, target - nums[i], i);
                curr.pop_back();
            }
        }
    }
};